### PR TITLE
Add field based sorting to the kubectl command line.

### PR DESCRIPTION
**Pull request by brendandburns**
Fixes https://github.com/GoogleCloudPlatform/kubernetes/issues/12402

@jlowdermilk @bgrant0607 @smarterclayton

### CODE REVIEW

File: pkg/kubectl/resource_printer.go [hunk 1]
```diff
@@ -1198,3 +1199,85 @@ func indirect(v reflect.Value) (rv reflect.Value, isNil bool) {
...
+func (r *RuntimeSort) Less(i, j int) bool {
+	iObj := r.objs[i]
+	jObj := r.objs[j]
+
+	parser := jsonpath.New("sorting")
+	parser.Parse(r.field)
+
+	iValues, err := parser.FindResults(reflect.ValueOf(iObj).Elem().Interface())
+	if err != nil {
+		glog.Fatalf("Failed to get i values for %#v using %s (%#v)", iObj, r.field, err)
```
Comment (jlowdermilk) [inline pkg/kubectl/resource_printer.go]: Can we put a check that the path exists for the object(s) at the beginning of `sortObj`, and surface an error if it doesn't? This is not a friendly error to get if a user misplaces a period or a capital in a fieldpath.

File: pkg/kubectl/resource_printer.go [hunk 2]
```diff
@@ -1198,3 +1199,85 @@ func indirect(v reflect.Value) (rv reflect.Value, isNil bool) {
 	}
 	return v, false
 }
+
```
Comment (smarterclayton) [inline pkg/kubectl/resource_printer.go]: Put this in its own file.  This file is already way too big.

**Comment by smarterclayton** [comment]
cc @kargakis

File: pkg/kubectl/resource_printer.go (hunk 1 above)
Comment (brendandburns) [inline pkg/kubectl/resource_printer.go]: Done.

File: pkg/kubectl/resource_printer.go (hunk 2 above)
Comment (brendandburns) [inline pkg/kubectl/resource_printer.go]: done.

**Comment by brendandburns** [comment]
Comments addressed, please re-check.

File: pkg/kubectl/cmd/util/printing.go [hunk 3]
```diff
@@ -32,6 +32,7 @@ func AddPrinterFlags(cmd *cobra.Command) {
 	cmd.Flags().String("output-version", "", "Output the formatted object with the given version (default api-version).")
 	cmd.Flags().Bool("no-headers", false, "When using the default output, don't print headers.")
 	cmd.Flags().StringP("template", "t", "", "Template string or path to template file to use when -o=template or -o=templatefile.  The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview]")
+	cmd.Flags().String("sort-by", "", "If non-empty, sort list types using this field.  The field must be an integer or a string, the template format is JSONPath, (e.g. '.ObjectMeta.Name')")
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/util/printing.go]: This actually only supports jsonpath at the moment (and not integer column numbers), doesn't it?

Comment (brendandburns) [inline pkg/kubectl/cmd/util/printing.go]: This comment was a little vague.  The field specification itself (the JSONPath thing) needs to be a string, but the field in the resource that it points to can be a string or int.

I clarified the wording to make this apparent.

**Comment by brendandburns** [comment]
Comment addressed, ptal.

--brendan

File: pkg/kubectl/sorting_printer.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		return nil
+	}
+	parser := jsonpath.New("sorting")
+	parser.Parse(s.SortField)
+	values, err := parser.FindResults(reflect.ValueOf(objs[0]).Elem().Interface())
+	if err != nil {
+		return err
+	}
+	if len(values) == 0 {
+		return fmt.Errorf("Couldn't find any field with path: %s", s.SortField)
```
Comment (jlowdermilk) [inline pkg/kubectl/sorting_printer.go]: Nit: lowercase for returned error message.

**Comment by bgrant0607** [comment]
The new text is more clear. Thanks.

**Comment by bgrant0607** [comment]
cc @JanetKuo

**Comment by bgrant0607** [comment]
Needs to be rebased, apparently.

File: pkg/kubectl/sorting_printer.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	"sort"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/jsonpath"
+)
+
+// Sorting printer sorts list types before delegating to another printer.
+// Non-list types are simply passed through
+type SortingPrinter struct {
```
Comment (kargakis) [inline pkg/kubectl/sorting_printer.go]: Can you move this above its methods?

File: pkg/kubectl/sorting_printer_test.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+					},
+				},
+			},
+			field: "{.Spec.Replicas}",
+		},
+	}
+	for _, test := range tests {
+		sort := &SortingPrinter{SortField: test.field}
+		if err := sort.sortObj(test.obj); err != nil {
+			t.Errorf("unexpected error: %v", err)
```
Comment (kargakis) [inline pkg/kubectl/sorting_printer_test.go]: Add a name field to the tests struct and pass it here like "%s: unexpected error" so we know which test fails if/when they fail. Also if you fail here you don't need to continue to `DeepEqual`.

**Comment by kargakis** [comment]
`expose`, `run`, and `rolling-update` won't have any use for `sort-by` since they all print single objects.

**Comment by bgrant0607** [comment]
Mutations in general won't need sort-by.

File: pkg/kubectl/sorting_printer.go (hunk 5 above)
Comment (brendandburns) [inline pkg/kubectl/sorting_printer.go]: done (moved its methods up)

File: pkg/kubectl/sorting_printer_test.go (hunk 6 above)
Comment (brendandburns) [inline pkg/kubectl/sorting_printer_test.go]: done.

**Comment by brendandburns** [comment]
Regarding it being unnecessary, I'd rather leave the printing stuff in a single location for all usages, it's a no-op if the resource isn't a List anyway.

ptal.

--brendan

**Comment by jlowdermilk** [comment]
Reassigning to @kargakis, since he has remaining comments.

File: pkg/kubectl/sorting_printer_test.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+		},
+	}
+	for _, test := range tests {
+		sort := &SortingPrinter{SortField: test.field}
+		if err := sort.sortObj(test.obj); err != nil {
+			t.Errorf("unexpected error: %v (%s)", err, test.name)
+			continue
+		}
+		if !reflect.DeepEqual(test.obj, test.sort) {
+			t.Errorf("expected:\n%v\nsaw:\n%v", test.sort, test.obj)
```
Comment (kargakis) [inline pkg/kubectl/sorting_printer_test.go]: Add the test name here too

**Comment by kargakis** [comment]
One more comment and this LGTM

File: pkg/kubectl/sorting_printer_test.go (hunk 7 above)
Comment (brendandburns) [inline pkg/kubectl/sorting_printer_test.go]: done

**Comment by brendandburns** [comment]
Comment addressed.  ptal.

**Comment by jlowdermilk** [comment]
What's with the 4000+ red lines of autogenerated deep copy functions? That wasn't there before...

**Comment by brendandburns** [comment]
Yeah, that's fixed.  The precommits crashed and in crashing they deleted
the deep copy code gen.  I reverted that.

Sorry!
Brendan

**Comment by jlowdermilk** [comment]
lgtm

**Pull request merged**